app/src/lib/editors/win32.ts

```diff
@@ -10,663 +10,595 @@ import {
+      },
+      // 32-bit version of Notepad++
+      {
+        key: HKEY.HKEY_LOCAL_MACHINE,
+        subKey:
+          'SOFTWARE\\WOW6432Node\\Microsoft\\Windows\\CurrentVersion\\Uninstall\\Notepad++',
+      },
+    ],
+    executableShimPath: [],
+    appInformationExtractor: keys => {
+      const displayName = getKeyOrEmpty(keys, 'DisplayName')
+      const publisher = getKeyOrEmpty(keys, 'Publisher')
+      const installLocation = getKeyOrEmpty(keys, 'DisplayIcon')
+
+      return { displayName, publisher, installLocation }
+    },
+    expectedInstallationChecker: (displayName, publisher) =>
+      displayName.startsWith('Notepad++') && publisher === 'Notepad++ Team',
+  },
+  {
+    name: 'JetBrains Rider',
+    registryKeys: [
+      // Rider 2019.3.4
+      {
+        key: HKEY.HKEY_LOCAL_MACHINE,
+        subKey:
+          'SOFTWARE\\WOW6432Node\\Microsoft\\Windows\\CurrentVersion\\Uninstall\\JetBrains Rider 2019.3.4',
+      },
+    ],
+    executableShimPath: ['bin', 'rider64.exe'],
+    appInformationExtractor: keys => {
+      let displayName = ''
+      let publisher = ''
+      let installLocation = ''
+
+      for (const item of keys) {
+        // NOTE:
+        // JetBrains Rider adds the current release number to the end of the Display Name, below checks for "JetBrains Rider"
+        if (
+          item.name === 'DisplayName' &&
+          item.type === RegistryValueType.REG_SZ &&
+          item.data.startsWith('JetBrains Rider ')
+        ) {
+          displayName = 'JetBrains Rider'
+        } else if (
+          item.name === 'Publisher' &&
+          item.type === RegistryValueType.REG_SZ
+        ) {
+          publisher = item.data
+        } else if (
+          item.name === 'InstallLocation' &&
+          item.type === RegistryValueType.REG_SZ
+        ) {
+          installLocation = item.data
+        }
       }
-    }
 
-    return { displayName, publisher, installLocation }
-  }
+      return { displayName, publisher, installLocation }
+    },
+    expectedInstallationChecker: (displayName, publisher) =>
+      displayName.startsWith('JetBrains Rider') &&
+      publisher === 'JetBrains s.r.o.',
+  },
+]
 
-  return assertNever(editor, `Unknown external editor: ${editor}`)
+function getKeyOrEmpty(
+  keys: ReadonlyArray<RegistryValue>,
+  key: string
+): string {
+  const entry = keys.find(k => k.name === key)
+  return entry && entry.type === RegistryValueType.REG_SZ ? entry.data : ''
 }
 
-async function findApplication(editor: ExternalEditor): Promise<string | null> {
-  const registryKeys = getRegistryKeys(editor)
+const defaultAppInformationExtractor: AppInformationExtractor = keys => {
+  const displayName = getKeyOrEmpty(keys, 'DisplayName')
+  const publisher = getKeyOrEmpty(keys, 'Publisher')
+  const installLocation = getKeyOrEmpty(keys, 'InstallLocation')
+  return { displayName, publisher, installLocation }
+}
 
+async function findApplication(
```

Comment:
I know this isn't part of your refactor but upon reading this method again it sure feels weird that we're just grabbing the first set of keys and then not continuing down the fallback chain if the `expectedInstallationChecker`  fails.

Seems like this method should be rewritten such that we test the validity of each set of keys. Something like this:

```ts
async function findApplication(editor: IWindowsExternalEditor) {
  const getAppInfo =
    editor.appInformationExtractor ?? defaultAppInformationExtractor

  for (const { key, subKey } of editor.registryKeys) {
    const keys = enumerateValues(key, subKey)
    if (keys.length === 0) {
      continue
    }

    const { displayName, publisher, installLocation } = getAppInfo(keys)

    if (!editor.expectedInstallationChecker(displayName, publisher)) {
      log.debug(`Unexpectted registry entries for ${editor.name}`)
      continue
    }

    const path = Path.join(installLocation, ...editor.executableShimPath)
    const exists = await pathExists(path)
    if (!exists) {
      log.debug(`Executable for ${editor.name} not found at '${path}'`)
      continue
    }

    return path
  }

  return null
}
```